File: solutions/javascript/list-ops/list-ops.js
// @ts-check
export class List {
  constructor(/** @type {any[] | undefined} list */ list) {
    this.values = list || [];
  }

  /**
   * Given a list, add all items in the provided list to the end of this list
   *
   * @param {List} list
   * @return {List} appended values
   */
  append(list) {
    let appended = this.values;
    for (const item of list.values) {
      appended = [...appended, item];
    }
    return new List(appended);
  }

  /**
   * Given a list of lists, combine all items in all lists into one flattened list
   *
   * @param {List} lists
   * @return {List} concatenated values
   */
  concat(lists) {
    let concatenatedList = new List(this.values);
    for (const list of lists.values) {
      concatenatedList = concatenatedList.append(list);
    }
    return concatenatedList;
  }

  /**
   * Given a predicate, return the list of all items for which predicate(item) is True
   *
   * @param {(item: any) => boolean} pred
   * @return {List} filtered values
   */
  filter(pred) {
    let filtered = [];
    for (const item of this.values) {
      filtered = pred(item) ? [...filtered, item] : filtered;
    }
    return new List(filtered);
  }

  /**
   * Given a function, return the list of the results of applying function(item) on all items
   *
   * @param {(item: any) => any} func
   * @return {List} mapped values
   */
  map(func) {
    let mapped = [];
    for (const item of this.values) {
      mapped = [...mapped, func(item)];
    }
    return new List(mapped);
  }

  /**
   * Return the total number of items within it
   *
   * @return {number} count of values
   */
  length() {
    let count = 0;
    for (const _ of this.values) {
      count += 1;
    }
    return count;
  }

  /**
   * Given a function and initial accumulator, fold each item into the accumulator from the left
   *
   * @param {(acc: any, item: any) => any} func
   * @param {any} initial
   * @return {any} accumulated value
   */
  foldl(func, initial) {
    let final = initial;
    for (const item of this.values) {
      final = func(final, item);
    }
    return final;
  }

  /**
   * Given a function and initial accumulator, fold each item into the accumulator from the right
   *
   * @param {(acc: any, item: any) => any} func
   * @param {any} initial
   * @return {any} accumulated value
   */
  foldr(func, initial) {
    let final = initial;
    for (const item of this.reverse().values) {
      final = func(final, item);
    }
    return final;
  }

  /**
   * Return a list with all the original items, but in reversed order
   *
   * @return {List} reversed values
   */
  reverse() {
    let reversed = [];
    for (const item of this.values) {
      reversed = [item, ...reversed];
    }
    return new List(reversed);
  }
}

File: solutions/go/list-ops/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	result := initial
	for _, item := range s {
		result = fn(result, item)
	}
	return result
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	result := initial
	for i := range s {
		result = fn(s[len(s)-1-i], result)
	}
	return result
}

func (s IntList) Filter(fn func(int) bool) IntList {
	filtered := IntList{}
	for _, item := range s {
		if fn(item) {
			filtered = append(filtered, item)
		}
	}
	return filtered
}

func (s IntList) Length() int {
	return s.Foldl(func(count, _ int) int { return count + 1 }, 0)
}

func (s IntList) Map(fn func(int) int) IntList {
	mapped := IntList{}
	for _, item := range s {
		mapped = append(mapped, fn(item))
	}
	return mapped
}

func (s IntList) Reverse() IntList {
	reversed := IntList{}
	for i := range s {
		reversed = append(reversed, s[len(s)-1-i])
	}
	return reversed
}

func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

func (s IntList) Concat(lists []IntList) IntList {
	concatenated := s
	for _, list := range lists {
		concatenated = concatenated.Append(list)
	}
	return concatenated
}
